simulator/failures/utils.py

```diff
@@ -0,0 +1,344 @@
+            data, target = data.cuda(), target.cuda()
+            output = global_model(data)
+            loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
+            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
+            correct += pred.eq(target.view_as(pred)).sum().item()
+
+    loss /= len(data_loader.dataset)
+    acc = correct / len(data_loader.dataset)
+
+    return loss, acc
+
+def consensus(global_model, client_models):
+    """
+    Computes the consensus value of each client model by comparing it to the global
+    one. We use Mean Square Error to measure the distance between two weight vectors
+    and average the consensus values of each parameter vector.
+
+    Params:
+        global_model (nn.Module): the global neural network averaging all the clients
+        client_models (array of Net): the list of all the client neural networks
+
+    Returns:
+        cons(array): the list of consensus values, one per client
+    """
+    cos = nn.CosineSimilarity(dim=0, eps=1e-6)
+    global_dict_values = list(global_model.state_dict().values())
+    n = len(global_dict_values)
+    cons = []
+    for client in client_models:
+        k = 0
+        client_dict_values = list(client.state_dict().values())
+        dist = [torch.norm(client_dict_values[k] - global_dict_values[k]) ** 2 for k in range(n)]
+        mean_dist = torch.mean(torch.Tensor(dist))
+        cons.append(round(mean_dist.item(), 6))
+    return cons
+
+def load_data(batch_size, num_clients, distribution='iid'):
+    """
+    Loads data and spreads it across the clients.
+
+    Params:
+        distribution (string): how the data is distributed across clients, independently and identically by default
+        batch_size (int): the chosen data batch size
+        num_clients (int): the number of clients in the network to spread the data on
+
+    Returns:
+        train_loader (array): the list of all train datasets, one per client
+        test_loader (array): the list of test datasets, one per client
+    """
+    assert distribution in ['iid', 'non-iid']
+
+    if distribution == 'iid':
+        traindata = datasets.MNIST('./data', train=True, download=True,
+                       transform=transforms.Compose([transforms.ToTensor(),transforms.Normalize((0.1307,), (0.3081,))])
+                       )
+        traindata_split = torch.utils.data.random_split(traindata, [int(traindata.data.shape[0] / num_clients) for _ in range(num_clients)])
+        train_loader = [torch.utils.data.DataLoader(x, batch_size=batch_size, shuffle=True) for x in traindata_split]
+
+        test_loader = torch.utils.data.DataLoader(
+                datasets.MNIST('./data', train=False, transform=transforms.Compose([transforms.ToTensor(),transforms.Normalize((0.1307,), (0.3081,))])
+                ), batch_size=batch_size, shuffle=True)
+    else: # distribution == 'non-iid'
```

Comment:
we may include more description details and other degrees of non-iid-ness for the `non-iid` data distribution.